src/ddsrt/src/process/posix/process.c

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright(c) 2006 to 2018 ADLINK Technology Limited and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+#include <unistd.h>
+#include <assert.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include "dds/ddsrt/process.h"
+
+
+ddsrt_pid_t
+ddsrt_getpid(void)
+{
+  /* Mapped to taskIdSelf() in VxWorks kernel mode. */
+  return getpid();
+}
+
+
+#ifdef DDSRT_USE_PROCESSCREATION
+
+#ifndef PIKEOS_POSIX
+#include <sys/wait.h>
+#endif
+
+#if defined(__APPLE__)
+#include <crt_externs.h>
+#else
+/* environ is a variable declared in unistd.h, and it keeps track
+ * of the environment variables during this running process. */
+extern char **environ;
+#endif
+
+#include "dds/ddsrt/heap.h"
+#include "dds/ddsrt/string.h"
+
+
+
+/*
+ * This'll take a argv and prefixes it with the given prefix.
+ * If argv is NULL, the new argv array will only contain the prefix and a NULL.
+ * The result array is always terminated with NULL.
+ */
+static char**
+prefix_argv(const char *prefix, char *const argv_in[])
+{
+  char **argv_out;
+
+  assert(prefix);
+
+  if (argv_in == NULL) {
+    argv_out = ddsrt_malloc(2 * sizeof(char*));
+    argv_out[0] = (char*)prefix;
+    argv_out[1] = NULL;
+  } else {
+    size_t argi;
+    size_t argc = 0;
+    while (argv_in[argc] != NULL) {
+      argc++;
+    }
+    argv_out = ddsrt_malloc((argc + 2) * sizeof(char*));
+    argv_out[0] = (char*)prefix;
+    for (argi = 0; argi < argc; argi++) {
+      argv_out[argi + 1] = (char*)argv_in[argi];
+    }
+    argv_out[argc + 1] = NULL;
+  }
+
+  return argv_out;
+}
+
+
+
+dds_retcode_t
+ddsrt_process_create(
+  const char *executable,
+  char *const argv[],
+  ddsrt_pid_t *pid)
+{
+  dds_retcode_t rv = DDS_RETCODE_OK;
+  pid_t spawn;
+
+  assert(executable != NULL);
+  assert(pid != NULL);
+
+  /* Does it exist and is it actually an executable? */
+  if (access(executable, X_OK) != 0) {
```

Comment:
Agree, but actually this test doesn't add anything of value anyway. The test may succeed, but the file may be gone (or otherwise changed) before the call to execve. So, it would be much better to rely on the return value of execve and use the same conventions as, e.g., bash to indicate problems:

> If a command is not found, the child process created to execute it returns a status of 127.  If a command is found but is not executable, the return status is 126.

Reply:
@eboasson How do I (easily) get the execv return value (when it fails) in the parent process to use that as return value for the create function (I don't wonna wait)? As far as I know, that's not possible. Am I missing something?

Reply:
You can't easily do that ... you could build a relatively complicated machinery with pipes and close-on-exec, but you could also simply use the exit status of the child and reserve a range of exit codes for this purpose (like bash effectively does). For this purpose it would probably be good enough.

Reply:
Let me get this straight. This would mean that ddsrt_process_create() always returns 'ok' (assuming the fork always succeeds). Only after I call a waitpid, do I know that the actual creation failed. Is this summary correct?

This would be quite different compared to how it works on Windows and we'd have different behaviours for ddsrt_process_create().

Reply:
Fair enough — a close-on-exec pipe is indeed the better option in which you write the error code (if any). That way, you either get end-of-file or an error code.

Reply:
Added the close-on-exec pipe and improved return values.